plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -423,11 +484,352 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
+	potentialVictims := util.SortableList{CompFunc: util.HigherPriorityPod}
+	nodeInfoCopy := nodeInfo.Clone()
+
+	removePod := func(rp *v1.Pod) {
+		nodeInfoCopy.RemovePod(rp)
+		if meta != nil {
+			meta.RemovePod(rp)
+		}
+	}
+	addPod := func(ap *v1.Pod) {
+		nodeInfoCopy.AddPod(ap)
+		if meta != nil {
+			meta.AddPod(ap, nodeInfoCopy)
+		}
+	}
+	// As the first step, remove all the lower priority pods from the node and
+	// check if the given pod can be scheduled.
+	podPriority := util.GetPodPriority(pod)
+	for _, p := range nodeInfoCopy.Pods() {
+		if util.GetPodPriority(p) < podPriority {
+			potentialVictims.Items = append(potentialVictims.Items, p)
+			removePod(p)
+		}
+	}
+	potentialVictims.Sort()
+	// If the new pod does not fit after removing all the lower priority pods,
+	// we are almost done and this node is not suitable for preemption. The only condition
+	// that we should check is if the "pod" is failing to schedule due to pod affinity
+	// failure.
+	if fits, failedPredicates, err := podFitsOnNode(pod, meta, nodeInfoCopy, fitPredicates, nil); !fits {
+		if err != nil {
+			glog.Warningf("Encountered error while selecting victims on node %v: %v", nodeInfo.Node().Name, err)
+			return nil, false
+		}
+		// If the new pod still cannot be scheduled for any reason other than pod
+		// affinity, the new pod will not fit on this node and we are done here.
+		affinity := pod.Spec.Affinity
+		if affinity == nil || affinity.PodAffinity == nil {
+			return nil, false
+		}
+		for _, failedPred := range failedPredicates {
+			if failedPred != predicates.ErrPodAffinityNotMatch {
+				return nil, false
+			}
+		}
+		// If we reach here, it means that the pod cannot be scheduled due to pod
+		// affinity or anti-affinity. Since failure reason for both affinity and
+		// anti-affinity is the same, we cannot say which one caused it. So, we try
+		// adding pods one at a time and see if any of them satisfies the affinity rules.
+		interPodAffinityPredicate, found := fitPredicates[predicates.MatchInterPodAffinity]
+		if !found {
+			glog.Warning("Unable to find inter-pod affinity predicate.")
+			return nil, false
+		}
+		// We add all the pods needed to satisfy the affinity rules first and then
+		// check if the preemptor still fits. If it does not, the preemption is not
+		// going to help the pod get scheduled on this node.
+		podIndexesForAffinity := []int{} // indexes of victims needed for affinity
+		affinityCanBeSatisfied := false
+		for i, p := range potentialVictims.Items {
+			existingPod := p.(*v1.Pod)
+			addPod(existingPod)
+			podIndexesForAffinity = append(podIndexesForAffinity, i)
+			if fits, _, _ = interPodAffinityPredicate(pod, meta, nodeInfoCopy); fits {
+				affinityCanBeSatisfied = true
+				break
+			}
+		}
+		if !affinityCanBeSatisfied {
+			// Even after adding all the victims, affinity rules cannot be satisfied.
+			// The failure may be related to anti-affinity to higher priority pods or
```

Comment:
What about anti-affinity? Are we going to handle it somehow differrently?

If there is a low priority pod on a different node that is blocking placing our pod on this node due to anti-affinity, it should be preempted. We are not even trying to solve this problem.

@davidopp

Reply:
@davidopp and I decided not to do cross node preemption at least in the first version due to its complications.

Reply:
@bsalamat I think it would be useful to write a comment (maybe start it as a Google Doc, and then move it into a comment) explaining what is the intended behavior for each of the possibilities. I believe "each of the possibilities" means
* new pod is initially infeasible due to new pod having pod affinity TO lower-priority pod on the same node
* new pod is initially infeasible due to new pod having pod anti-affinity TO lower-priority pod on the same node
* new pod is initially infeasible due to new pod having pod affinity FROM lower-priority pod on the same node
* new pod is initially infeasible due to new pod having pod anti-affinity FROM lower-priority pod on the same node
and then all four of the above but for "pod on another node"
(so a total of eight entries)

Does that sound reasonable?

Reply:
I updated my comment; please read on the web if you initially read it by email.

Reply:
That is reasonable, but before that I would like to ask what do you think if we say that we don't support affinity on lower-priority pods during preemption, i.e., preemption logic may or may not preempt other pods in this case, so users shouldn't rely on it. I have explained in other comments why we should not commit to support it.